scratch_space/lib/src/util.dart

```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:path/path.dart' as p;
+
+/// Returns the top level directory in [uri].
+///
+/// Throws an [ArgumentError] if [uri] is just a filename with no directory.
+String topLevelDir(String uri) {
+  var parts = p.url.split(p.url.normalize(uri));
+  String error;
+  if (parts.length == 1) {
+    error = 'The uri `$uri` does not contain a directory.';
+  } else if (parts.first == '..') {
```

Comment:
This would also escape: `/allowed/path/../../../not_allowed`

I think we need `parts.contains('..')`

Reply:
There are actually tests for this case :). Since we normalize the uri first we only have to check the first segment.

Reply:
I wasn't actually testing the basic case so I added a test for that as well